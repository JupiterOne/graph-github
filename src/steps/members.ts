import {
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
  IntegrationMissingKeyError,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../client';
import { IntegrationConfig } from '../config';
import { DATA_ACCOUNT_ENTITY } from './account';
import {
  toOrganizationMemberEntity,
  toOrganizationHasMemberRelationship,
  toMemberManagesOrganizationRelationship,
} from '../sync/converters';
import { AccountEntity, UserEntity, IdEntityMap } from '../types';
import { OrgMemberRole } from '../client/GraphQLClient';
import {
  GITHUB_ACCOUNT_ENTITY_TYPE,
  GITHUB_MEMBER_ENTITY_TYPE,
  GITHUB_MEMBER_ENTITY_CLASS,
  GITHUB_MEMBER_ACCOUNT_RELATIONSHIP_TYPE,
  GITHUB_ACCOUNT_MEMBER_RELATIONSHIP_TYPE,
} from '../constants';

export async function fetchMembers({
  instance,
  logger,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const config = instance.config;
  const apiClient = createAPIClient(config, logger);

  const accountEntity = (await jobState.getData<AccountEntity>(
    DATA_ACCOUNT_ENTITY,
  )) as AccountEntity;
  if (!accountEntity) {
    throw new IntegrationMissingKeyError(
      `Expected to find Account entity in jobState.`,
    );
  }

  //for use later in PRs
  const memberEntities: UserEntity[] = [];
  const memberByLoginMap: IdEntityMap<UserEntity> = {};

  await apiClient.iterateMembers(async (member) => {
    const memberEntity = (await jobState.addEntity(
      toOrganizationMemberEntity(member),
    )) as UserEntity;

    memberEntities.push(memberEntity);
    memberByLoginMap[member.login] = memberEntity;

    await jobState.addRelationship(
      toOrganizationHasMemberRelationship(accountEntity, memberEntity),
    );

    if (memberEntity.role === OrgMemberRole.Admin) {
      await jobState.addRelationship(
        toMemberManagesOrganizationRelationship(accountEntity, memberEntity),
      );
    }
  });

  await jobState.setData('MEMBER_ARRAY', memberEntities);
  await jobState.setData('MEMBER_BY_LOGIN_MAP', memberByLoginMap);
}

export const memberSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-users',
    name: 'Fetch Users',
    entities: [
      {
        resourceName: 'Github User',
        _type: GITHUB_MEMBER_ENTITY_TYPE,
        _class: GITHUB_MEMBER_ENTITY_CLASS,
      },
    ],
    relationships: [
      {
        _type: GITHUB_ACCOUNT_MEMBER_RELATIONSHIP_TYPE,
        _class: RelationshipClass.HAS,
        sourceType: GITHUB_ACCOUNT_ENTITY_TYPE,
        targetType: GITHUB_MEMBER_ENTITY_TYPE,
      },
      {
        _type: GITHUB_MEMBER_ACCOUNT_RELATIONSHIP_TYPE,
        _class: RelationshipClass.MANAGES,
        sourceType: GITHUB_MEMBER_ENTITY_TYPE,
        targetType: GITHUB_ACCOUNT_ENTITY_TYPE,
      },
    ],
    dependsOn: ['fetch-account'],
    executionHandler: fetchMembers,
  },
];
