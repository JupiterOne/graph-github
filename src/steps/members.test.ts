import { Recording } from '@jupiterone/integration-sdk-testing';
import { sanitizeConfig } from '../config';
import { memberSteps } from './members';
import { integrationConfig } from '../../test/config';
import { setupGithubRecording } from '../../test/recording';
import {
  GithubEntities,
  GITHUB_MEMBER_BY_LOGIN_MAP,
  Relationships,
} from '../constants';
import { invocationConfig } from '..';
import { executeStepWithDependencies } from '../../test/executeStepWithDependencies';

jest.setTimeout(20000);

let recording: Recording;
afterEach(async () => {
  await recording.stop();
});

test('fetchMembers exec handler', async () => {
  recording = setupGithubRecording({
    directory: __dirname,
    name: 'members',
  });
  sanitizeConfig(integrationConfig);

  const {
    collectedEntities,
    collectedRelationships,
    encounteredTypes,
    jobState,
  } = await executeStepWithDependencies({
    stepId: memberSteps[0].id,
    invocationConfig: invocationConfig as any,
    instanceConfig: integrationConfig,
  });

  expect({
    numCollectedEntities: collectedEntities.length,
    numCollectedRelationships: collectedRelationships.length,
    collectedEntities: collectedEntities,
    collectedRelationships: collectedRelationships,
    encounteredTypes: encounteredTypes,
  }).toMatchSnapshot();

  const members = collectedEntities.filter(
    (e) => e._type === GithubEntities.GITHUB_MEMBER._type,
  );
  expect(members.length).toBeGreaterThan(0);
  expect(members).toMatchGraphObjectSchema(GithubEntities.GITHUB_MEMBER);

  // relationships
  const accountHasUserRels = collectedRelationships.filter(
    (e) => e._type === Relationships.ACCOUNT_HAS_USER._type,
  );
  expect(accountHasUserRels.length).toBeGreaterThan(0);

  const userManagesAccountRels = collectedRelationships.filter(
    (e) => e._type === Relationships.USER_MANAGES_ACCOUNT._type,
  );
  expect(userManagesAccountRels.length).toBeGreaterThan(0);

  // ensure that we are setting the GITHUB_MEMBER_BY_LOGIN_MAP in the jobState as expected
  const memberByLoginMap = await jobState.getData(GITHUB_MEMBER_BY_LOGIN_MAP);
  expect(memberByLoginMap).toBeTruthy();
});
