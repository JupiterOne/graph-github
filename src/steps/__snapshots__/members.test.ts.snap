// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchMembers exec handler 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "MDQ6VXNlcjEyODM4ODQ=",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "company": null,
            "createdAt": "2011-12-24T16:23:19Z",
            "databaseId": 1283884,
            "email": "sivli.embir@jupiterone.com",
            "hasTwoFactorEnabled": null,
            "id": "MDQ6VXNlcjEyODM4ODQ=",
            "isEmployee": false,
            "isSiteAdmin": false,
            "location": "Portland OR",
            "login": "Sivli-Embir",
            "name": "Sivli Embir",
            "organization": "O_kgDOB0VI3Q",
            "role": "ADMIN",
            "updatedAt": "2023-07-17T19:54:11Z",
            "url": "https://github.com/Sivli-Embir",
            "websiteUrl": null,
          },
        },
      ],
      "_type": "github_user",
      "active": true,
      "company": "",
      "createdOn": 1324743799000,
      "databaseId": 1283884,
      "displayName": "Sivli Embir",
      "email": "sivli.embir@jupiterone.com",
      "employee": false,
      "id": "MDQ6VXNlcjEyODM4ODQ=",
      "location": "Portland OR",
      "login": "Sivli-Embir",
      "mfaEnabled": undefined,
      "name": "Sivli Embir",
      "node": "MDQ6VXNlcjEyODM4ODQ=",
      "organizationId": "O_kgDOB0VI3Q",
      "role": "ADMIN",
      "siteAdmin": false,
      "updatedOn": 1689623651000,
      "username": "Sivli-Embir",
      "webLink": "https://github.com/Sivli-Embir",
      "websiteUrl": "",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "MDQ6VXNlcjIzMzYyNjI=",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "company": "JupiterOne",
            "createdAt": "2012-09-13T02:33:25Z",
            "databaseId": 2336262,
            "email": "ryan.mcafee@jupiterone.com",
            "hasTwoFactorEnabled": null,
            "id": "MDQ6VXNlcjIzMzYyNjI=",
            "isEmployee": false,
            "isSiteAdmin": false,
            "location": "Pensacola, Florida",
            "login": "ryanmcafee",
            "name": "Ryan McAfee",
            "organization": "O_kgDOB0VI3Q",
            "role": "ADMIN",
            "updatedAt": "2023-02-09T17:10:58Z",
            "url": "https://github.com/ryanmcafee",
            "websiteUrl": "https://ryanmcafee.com",
          },
        },
      ],
      "_type": "github_user",
      "active": true,
      "company": "JupiterOne",
      "createdOn": 1347503605000,
      "databaseId": 2336262,
      "displayName": "Ryan McAfee",
      "email": "ryan.mcafee@jupiterone.com",
      "employee": false,
      "id": "MDQ6VXNlcjIzMzYyNjI=",
      "location": "Pensacola, Florida",
      "login": "ryanmcafee",
      "mfaEnabled": undefined,
      "name": "Ryan McAfee",
      "node": "MDQ6VXNlcjIzMzYyNjI=",
      "organizationId": "O_kgDOB0VI3Q",
      "role": "ADMIN",
      "siteAdmin": false,
      "updatedOn": 1675962658000,
      "username": "ryanmcafee",
      "webLink": "https://github.com/ryanmcafee",
      "websiteUrl": "https://ryanmcafee.com",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "MDQ6VXNlcjM3NzE5MjQ=",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "company": "@JupiterOne ",
            "createdAt": "2013-03-05T03:49:20Z",
            "databaseId": 3771924,
            "email": "a@alk.im",
            "hasTwoFactorEnabled": null,
            "id": "MDQ6VXNlcjM3NzE5MjQ=",
            "isEmployee": false,
            "isSiteAdmin": false,
            "location": "Raleigh-Durham, North Carolina",
            "login": "austinkelleher",
            "name": "Austin Kelleher",
            "organization": "O_kgDOB0VI3Q",
            "role": "ADMIN",
            "updatedAt": "2023-06-16T18:01:43Z",
            "url": "https://github.com/austinkelleher",
            "websiteUrl": "https://www.austinkelleher.com",
          },
        },
      ],
      "_type": "github_user",
      "active": true,
      "company": "@JupiterOne ",
      "createdOn": 1362455360000,
      "databaseId": 3771924,
      "displayName": "Austin Kelleher",
      "email": "a@alk.im",
      "employee": false,
      "id": "MDQ6VXNlcjM3NzE5MjQ=",
      "location": "Raleigh-Durham, North Carolina",
      "login": "austinkelleher",
      "mfaEnabled": undefined,
      "name": "Austin Kelleher",
      "node": "MDQ6VXNlcjM3NzE5MjQ=",
      "organizationId": "O_kgDOB0VI3Q",
      "role": "ADMIN",
      "siteAdmin": false,
      "updatedOn": 1686938503000,
      "username": "austinkelleher",
      "webLink": "https://github.com/austinkelleher",
      "websiteUrl": "https://www.austinkelleher.com",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "MDQ6VXNlcjU5MjMwMDI=",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "company": "JupiterOne",
            "createdAt": "2013-11-12T20:20:50Z",
            "databaseId": 5923002,
            "email": "samuel.poulton@jupiterone.com",
            "hasTwoFactorEnabled": null,
            "id": "MDQ6VXNlcjU5MjMwMDI=",
            "isEmployee": false,
            "isSiteAdmin": false,
            "location": null,
            "login": "VDubber",
            "name": "Samuel Poulton",
            "organization": "O_kgDOB0VI3Q",
            "role": "ADMIN",
            "updatedAt": "2023-07-17T23:18:53Z",
            "url": "https://github.com/VDubber",
            "websiteUrl": null,
          },
        },
      ],
      "_type": "github_user",
      "active": true,
      "company": "JupiterOne",
      "createdOn": 1384287650000,
      "databaseId": 5923002,
      "displayName": "Samuel Poulton",
      "email": "samuel.poulton@jupiterone.com",
      "employee": false,
      "id": "MDQ6VXNlcjU5MjMwMDI=",
      "location": "",
      "login": "VDubber",
      "mfaEnabled": undefined,
      "name": "Samuel Poulton",
      "node": "MDQ6VXNlcjU5MjMwMDI=",
      "organizationId": "O_kgDOB0VI3Q",
      "role": "ADMIN",
      "siteAdmin": false,
      "updatedOn": 1689635933000,
      "username": "VDubber",
      "webLink": "https://github.com/VDubber",
      "websiteUrl": "",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "MDQ6VXNlcjcxOTA2OTY=",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "company": null,
            "createdAt": "2014-04-05T17:31:25Z",
            "databaseId": 7190696,
            "email": "",
            "hasTwoFactorEnabled": null,
            "id": "MDQ6VXNlcjcxOTA2OTY=",
            "isEmployee": false,
            "isSiteAdmin": false,
            "location": null,
            "login": "RonaldEAM",
            "name": "Ronald Arias",
            "organization": "O_kgDOB0VI3Q",
            "role": "ADMIN",
            "updatedAt": "2023-07-10T19:53:28Z",
            "url": "https://github.com/RonaldEAM",
            "websiteUrl": null,
          },
        },
      ],
      "_type": "github_user",
      "active": true,
      "company": "",
      "createdOn": 1396719085000,
      "databaseId": 7190696,
      "displayName": "Ronald Arias",
      "employee": false,
      "id": "MDQ6VXNlcjcxOTA2OTY=",
      "location": "",
      "login": "RonaldEAM",
      "mfaEnabled": undefined,
      "name": "Ronald Arias",
      "node": "MDQ6VXNlcjcxOTA2OTY=",
      "organizationId": "O_kgDOB0VI3Q",
      "role": "ADMIN",
      "siteAdmin": false,
      "updatedOn": 1689018808000,
      "username": "RonaldEAM",
      "webLink": "https://github.com/RonaldEAM",
      "websiteUrl": "",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "MDQ6VXNlcjEwMDkwNjQ3",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "company": "JupiterOne",
            "createdAt": "2014-12-05T18:36:04Z",
            "databaseId": 10090647,
            "email": "kenneth.kaye@gmail.com",
            "hasTwoFactorEnabled": null,
            "id": "MDQ6VXNlcjEwMDkwNjQ3",
            "isEmployee": false,
            "isSiteAdmin": false,
            "location": "Colorado, US",
            "login": "behobu",
            "name": "Kenneth Kaye",
            "organization": "O_kgDOB0VI3Q",
            "role": "ADMIN",
            "updatedAt": "2023-06-30T23:53:01Z",
            "url": "https://github.com/behobu",
            "websiteUrl": null,
          },
        },
      ],
      "_type": "github_user",
      "active": true,
      "company": "JupiterOne",
      "createdOn": 1417804564000,
      "databaseId": 10090647,
      "displayName": "Kenneth Kaye",
      "email": "kenneth.kaye@gmail.com",
      "employee": false,
      "id": "MDQ6VXNlcjEwMDkwNjQ3",
      "location": "Colorado, US",
      "login": "behobu",
      "mfaEnabled": undefined,
      "name": "Kenneth Kaye",
      "node": "MDQ6VXNlcjEwMDkwNjQ3",
      "organizationId": "O_kgDOB0VI3Q",
      "role": "ADMIN",
      "siteAdmin": false,
      "updatedOn": 1688169181000,
      "username": "behobu",
      "webLink": "https://github.com/behobu",
      "websiteUrl": "",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "MDQ6VXNlcjE2NTYyMDg1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "company": null,
            "createdAt": "2016-01-05T16:36:27Z",
            "databaseId": 16562085,
            "email": "",
            "hasTwoFactorEnabled": null,
            "id": "MDQ6VXNlcjE2NTYyMDg1",
            "isEmployee": false,
            "isSiteAdmin": false,
            "location": "Raleigh, NC",
            "login": "electricgull",
            "name": "Cameron Griffin",
            "organization": "O_kgDOB0VI3Q",
            "role": "ADMIN",
            "updatedAt": "2023-07-13T11:25:40Z",
            "url": "https://github.com/electricgull",
            "websiteUrl": null,
          },
        },
      ],
      "_type": "github_user",
      "active": true,
      "company": "",
      "createdOn": 1452011787000,
      "databaseId": 16562085,
      "displayName": "Cameron Griffin",
      "employee": false,
      "id": "MDQ6VXNlcjE2NTYyMDg1",
      "location": "Raleigh, NC",
      "login": "electricgull",
      "mfaEnabled": undefined,
      "name": "Cameron Griffin",
      "node": "MDQ6VXNlcjE2NTYyMDg1",
      "organizationId": "O_kgDOB0VI3Q",
      "role": "ADMIN",
      "siteAdmin": false,
      "updatedOn": 1689247540000,
      "username": "electricgull",
      "webLink": "https://github.com/electricgull",
      "websiteUrl": "",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "MDQ6VXNlcjMxMjE4NTA3",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "company": "@SignalPath",
            "createdAt": "2017-08-21T17:04:15Z",
            "databaseId": 31218507,
            "email": "",
            "hasTwoFactorEnabled": null,
            "id": "MDQ6VXNlcjMxMjE4NTA3",
            "isEmployee": false,
            "isSiteAdmin": false,
            "location": null,
            "login": "nealajpatel",
            "name": "Neal Patel",
            "organization": "O_kgDOB0VI3Q",
            "role": "ADMIN",
            "updatedAt": "2023-07-06T12:54:05Z",
            "url": "https://github.com/nealajpatel",
            "websiteUrl": null,
          },
        },
      ],
      "_type": "github_user",
      "active": true,
      "company": "@SignalPath",
      "createdOn": 1503335055000,
      "databaseId": 31218507,
      "displayName": "Neal Patel",
      "employee": false,
      "id": "MDQ6VXNlcjMxMjE4NTA3",
      "location": "",
      "login": "nealajpatel",
      "mfaEnabled": undefined,
      "name": "Neal Patel",
      "node": "MDQ6VXNlcjMxMjE4NTA3",
      "organizationId": "O_kgDOB0VI3Q",
      "role": "ADMIN",
      "siteAdmin": false,
      "updatedOn": 1688648045000,
      "username": "nealajpatel",
      "webLink": "https://github.com/nealajpatel",
      "websiteUrl": "",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "MDQ6VXNlcjU1NzcxODIz",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "company": null,
            "createdAt": "2019-09-25T00:55:40Z",
            "databaseId": 55771823,
            "email": "",
            "hasTwoFactorEnabled": null,
            "id": "MDQ6VXNlcjU1NzcxODIz",
            "isEmployee": false,
            "isSiteAdmin": false,
            "location": null,
            "login": "adam-in-ict",
            "name": "Adam Pierson",
            "organization": "O_kgDOB0VI3Q",
            "role": "ADMIN",
            "updatedAt": "2023-07-11T00:18:57Z",
            "url": "https://github.com/adam-in-ict",
            "websiteUrl": null,
          },
        },
      ],
      "_type": "github_user",
      "active": true,
      "company": "",
      "createdOn": 1569372940000,
      "databaseId": 55771823,
      "displayName": "Adam Pierson",
      "employee": false,
      "id": "MDQ6VXNlcjU1NzcxODIz",
      "location": "",
      "login": "adam-in-ict",
      "mfaEnabled": undefined,
      "name": "Adam Pierson",
      "node": "MDQ6VXNlcjU1NzcxODIz",
      "organizationId": "O_kgDOB0VI3Q",
      "role": "ADMIN",
      "siteAdmin": false,
      "updatedOn": 1689034737000,
      "username": "adam-in-ict",
      "webLink": "https://github.com/adam-in-ict",
      "websiteUrl": "",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "MDQ6VXNlcjczODM2MDA4",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "company": null,
            "createdAt": "2020-11-02T14:01:41Z",
            "databaseId": 73836008,
            "email": "",
            "hasTwoFactorEnabled": null,
            "id": "MDQ6VXNlcjczODM2MDA4",
            "isEmployee": false,
            "isSiteAdmin": false,
            "location": null,
            "login": "GuillaumeRoss",
            "name": "Guillaume Ross",
            "organization": "O_kgDOB0VI3Q",
            "role": "ADMIN",
            "updatedAt": "2023-07-20T21:41:12Z",
            "url": "https://github.com/GuillaumeRoss",
            "websiteUrl": null,
          },
        },
      ],
      "_type": "github_user",
      "active": true,
      "company": "",
      "createdOn": 1604325701000,
      "databaseId": 73836008,
      "displayName": "Guillaume Ross",
      "employee": false,
      "id": "MDQ6VXNlcjczODM2MDA4",
      "location": "",
      "login": "GuillaumeRoss",
      "mfaEnabled": undefined,
      "name": "Guillaume Ross",
      "node": "MDQ6VXNlcjczODM2MDA4",
      "organizationId": "O_kgDOB0VI3Q",
      "role": "ADMIN",
      "siteAdmin": false,
      "updatedOn": 1689889272000,
      "username": "GuillaumeRoss",
      "webLink": "https://github.com/GuillaumeRoss",
      "websiteUrl": "",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "U_kgDOBaVAgQ",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "company": null,
            "createdAt": "2021-11-19T19:46:17Z",
            "databaseId": 94716033,
            "email": "",
            "hasTwoFactorEnabled": null,
            "id": "U_kgDOBaVAgQ",
            "isEmployee": false,
            "isSiteAdmin": false,
            "location": null,
            "login": "j1-internal-automation",
            "name": "j1-internal-automation",
            "organization": "O_kgDOB0VI3Q",
            "role": "MEMBER",
            "updatedAt": "2023-06-16T18:12:11Z",
            "url": "https://github.com/j1-internal-automation",
            "websiteUrl": null,
          },
        },
      ],
      "_type": "github_user",
      "active": true,
      "company": "",
      "createdOn": 1637351177000,
      "databaseId": 94716033,
      "displayName": "j1-internal-automation",
      "employee": false,
      "id": "U_kgDOBaVAgQ",
      "location": "",
      "login": "j1-internal-automation",
      "mfaEnabled": undefined,
      "name": "j1-internal-automation",
      "node": "U_kgDOBaVAgQ",
      "organizationId": "O_kgDOB0VI3Q",
      "role": "MEMBER",
      "siteAdmin": false,
      "updatedOn": 1686939131000,
      "username": "j1-internal-automation",
      "webLink": "https://github.com/j1-internal-automation",
      "websiteUrl": "",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "U_kgDOBlNSJw",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "company": null,
            "createdAt": "2022-05-23T23:31:20Z",
            "databaseId": 106123815,
            "email": "",
            "hasTwoFactorEnabled": null,
            "id": "U_kgDOBlNSJw",
            "isEmployee": false,
            "isSiteAdmin": false,
            "location": null,
            "login": "QueenOvGraphs",
            "name": null,
            "organization": "O_kgDOB0VI3Q",
            "role": "MEMBER",
            "updatedAt": "2022-06-29T18:28:59Z",
            "url": "https://github.com/QueenOvGraphs",
            "websiteUrl": null,
          },
        },
      ],
      "_type": "github_user",
      "active": true,
      "company": "",
      "createdOn": 1653348680000,
      "databaseId": 106123815,
      "displayName": "QueenOvGraphs",
      "employee": false,
      "id": "U_kgDOBlNSJw",
      "location": "",
      "login": "QueenOvGraphs",
      "mfaEnabled": undefined,
      "name": "QueenOvGraphs",
      "node": "U_kgDOBlNSJw",
      "organizationId": "O_kgDOB0VI3Q",
      "role": "MEMBER",
      "siteAdmin": false,
      "updatedOn": 1656527339000,
      "username": "QueenOvGraphs",
      "webLink": "https://github.com/QueenOvGraphs",
      "websiteUrl": "",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "O_kgDOB0VI3Q",
      "_key": "O_kgDOB0VI3Q|has|MDQ6VXNlcjEyODM4ODQ=",
      "_toEntityKey": "MDQ6VXNlcjEyODM4ODQ=",
      "_type": "github_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjEyODM4ODQ=",
      "_key": "MDQ6VXNlcjEyODM4ODQ=|manages|O_kgDOB0VI3Q",
      "_toEntityKey": "O_kgDOB0VI3Q",
      "_type": "github_user_manages_account",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "O_kgDOB0VI3Q",
      "_key": "O_kgDOB0VI3Q|has|MDQ6VXNlcjIzMzYyNjI=",
      "_toEntityKey": "MDQ6VXNlcjIzMzYyNjI=",
      "_type": "github_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjIzMzYyNjI=",
      "_key": "MDQ6VXNlcjIzMzYyNjI=|manages|O_kgDOB0VI3Q",
      "_toEntityKey": "O_kgDOB0VI3Q",
      "_type": "github_user_manages_account",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "O_kgDOB0VI3Q",
      "_key": "O_kgDOB0VI3Q|has|MDQ6VXNlcjM3NzE5MjQ=",
      "_toEntityKey": "MDQ6VXNlcjM3NzE5MjQ=",
      "_type": "github_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjM3NzE5MjQ=",
      "_key": "MDQ6VXNlcjM3NzE5MjQ=|manages|O_kgDOB0VI3Q",
      "_toEntityKey": "O_kgDOB0VI3Q",
      "_type": "github_user_manages_account",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "O_kgDOB0VI3Q",
      "_key": "O_kgDOB0VI3Q|has|MDQ6VXNlcjU5MjMwMDI=",
      "_toEntityKey": "MDQ6VXNlcjU5MjMwMDI=",
      "_type": "github_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjU5MjMwMDI=",
      "_key": "MDQ6VXNlcjU5MjMwMDI=|manages|O_kgDOB0VI3Q",
      "_toEntityKey": "O_kgDOB0VI3Q",
      "_type": "github_user_manages_account",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "O_kgDOB0VI3Q",
      "_key": "O_kgDOB0VI3Q|has|MDQ6VXNlcjcxOTA2OTY=",
      "_toEntityKey": "MDQ6VXNlcjcxOTA2OTY=",
      "_type": "github_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjcxOTA2OTY=",
      "_key": "MDQ6VXNlcjcxOTA2OTY=|manages|O_kgDOB0VI3Q",
      "_toEntityKey": "O_kgDOB0VI3Q",
      "_type": "github_user_manages_account",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "O_kgDOB0VI3Q",
      "_key": "O_kgDOB0VI3Q|has|MDQ6VXNlcjEwMDkwNjQ3",
      "_toEntityKey": "MDQ6VXNlcjEwMDkwNjQ3",
      "_type": "github_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjEwMDkwNjQ3",
      "_key": "MDQ6VXNlcjEwMDkwNjQ3|manages|O_kgDOB0VI3Q",
      "_toEntityKey": "O_kgDOB0VI3Q",
      "_type": "github_user_manages_account",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "O_kgDOB0VI3Q",
      "_key": "O_kgDOB0VI3Q|has|MDQ6VXNlcjE2NTYyMDg1",
      "_toEntityKey": "MDQ6VXNlcjE2NTYyMDg1",
      "_type": "github_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjE2NTYyMDg1",
      "_key": "MDQ6VXNlcjE2NTYyMDg1|manages|O_kgDOB0VI3Q",
      "_toEntityKey": "O_kgDOB0VI3Q",
      "_type": "github_user_manages_account",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "O_kgDOB0VI3Q",
      "_key": "O_kgDOB0VI3Q|has|MDQ6VXNlcjMxMjE4NTA3",
      "_toEntityKey": "MDQ6VXNlcjMxMjE4NTA3",
      "_type": "github_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjMxMjE4NTA3",
      "_key": "MDQ6VXNlcjMxMjE4NTA3|manages|O_kgDOB0VI3Q",
      "_toEntityKey": "O_kgDOB0VI3Q",
      "_type": "github_user_manages_account",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "O_kgDOB0VI3Q",
      "_key": "O_kgDOB0VI3Q|has|MDQ6VXNlcjU1NzcxODIz",
      "_toEntityKey": "MDQ6VXNlcjU1NzcxODIz",
      "_type": "github_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjU1NzcxODIz",
      "_key": "MDQ6VXNlcjU1NzcxODIz|manages|O_kgDOB0VI3Q",
      "_toEntityKey": "O_kgDOB0VI3Q",
      "_type": "github_user_manages_account",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "O_kgDOB0VI3Q",
      "_key": "O_kgDOB0VI3Q|has|MDQ6VXNlcjczODM2MDA4",
      "_toEntityKey": "MDQ6VXNlcjczODM2MDA4",
      "_type": "github_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjczODM2MDA4",
      "_key": "MDQ6VXNlcjczODM2MDA4|manages|O_kgDOB0VI3Q",
      "_toEntityKey": "O_kgDOB0VI3Q",
      "_type": "github_user_manages_account",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "O_kgDOB0VI3Q",
      "_key": "O_kgDOB0VI3Q|has|U_kgDOBaVAgQ",
      "_toEntityKey": "U_kgDOBaVAgQ",
      "_type": "github_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "O_kgDOB0VI3Q",
      "_key": "O_kgDOB0VI3Q|has|U_kgDOBlNSJw",
      "_toEntityKey": "U_kgDOBlNSJw",
      "_type": "github_account_has_user",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "github_account",
    "github_user",
    "github_account_has_user",
    "github_user_manages_account",
  ],
  "numCollectedEntities": 12,
  "numCollectedRelationships": 22,
}
`;
